File: airial_api/database/object_models/project_management/herd_units.py
# Class definition for Herd Unit objects in the database
# Author: Michael B. Lance
#---------------------------------------------------------------------------------------------------------------------------#

from dataclasses import dataclass
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel

from ..base import DBbase

#---------------------------------------------------------------------------------------------------------------------------#

@dataclass
class HerdUnit(DBbase):
	''' Dataclass for representing Herd Units from the database
	
	'''
	herd_unit_id: int
	name: str
	created: datetime
	modified: datetime
	uuid: UUID

	#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#

	def to_dict(self):
		return {
			'herd_unit_id': self.herd_unit_id,
			'name': self.name,
			'created': self.fmt_date(self.created),
			'modified': self.fmt_date(self.modified),
			'uuid': str(self.uuid)
		}

#---------------------------------------------------------------------------------------------------------------------------#

class CreateHerdUnitReq(BaseModel):
	name: str
	project_id: UUID

File: generator_web_app/src/modules/api/herdunits.ts
// API methods for managing herd unit objects
// Author: Michael B. Lance

// ---------------------------------------------------------------------------------------------------------------------------

import { HerdUnit, Survey } from "@/types/generatorobjects.ts";
import type { HerdUnitIntf, SurveyIntf } from "@/types/generatorobjects.ts";
import { ApiError } from "@/modules/api/errors.ts";
import { api_url } from "@/modules/api/apiV1Methods.ts";

// ---------------------------------------------------------------------------------------------------------------------------

export async function getHerdUnitSurveys(
  herd_unit_id: string,
): Promise<Survey[]> {
  const response = await fetch(
    `${api_url}/herd-units/${herd_unit_id}/surveys`,
    {
      method: "GET",
      credentials: "include",
      headers: {
        "Content-Type": "application/json",
      },
    },
  );
  if (!response.ok) throw new ApiError(await response.json());
  const resp = await response.json();
  let surveys = [];
  for (const survey of resp) surveys.push(new Survey(survey as SurveyIntf));
  return surveys;
}

// ---------------------------------------------------------------------------------------------------------------------------

export interface createHerdUnitOptions {
  project_id: string;
  name: string
};

export async function createHerdUnit(
  options: createHerdUnitOptions,
): Promise<HerdUnit> {
  console.log(options);
  const response = await fetch(`${api_url}/herd-units`, {
    method: "POST",
    credentials: "include",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(options),
  });
  if (!response.ok) throw new ApiError(await response.json());

  return new HerdUnit((await response.json()) as HerdUnitIntf);
}

// ---------------------------------------------------------------------------------------------------------------------------
